accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+
+For .NET 8, there are two primary goals:
+1) Support byref-like types both for invoking and passing as arguments; this unblocks various scenarios. An unsafe approach may used for .NET 8 if support for `TypedReference` isn't addressed by Roslyn (covered later).
+2) Support "fast invoke" so that using IL Emit with dynamic methods has little to no performance advantage. Today both `STJ (System.Text.Json)` and [DI (dependency injection)](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL Emit for performance although DI uses emit through expressions while STJ uses IL Emit directly.
+    - Although the new APIs will be zero alloc (no `object` and boxing required) and performace faster than standard reflection, it will not necessarily be as fast as hand-coded IL Emit for specific scenarios that can optimize for any constraints such as fixed-size list of parameters or by not doing full defaulting and validation of values. Note that property get\set is a subset of this case since there is either a return value (for _get_) or a single parameter for _set_ and since property get\set is such a common case with serialization, this design does propose a separate API for this common case to enable maximum performance.
+
+# Design with managed pointers
+In order to address the limitations and issues, a least-common-denominator approach of using _managed pointers_ for the parameters, target and return value. This supports `in\ref\out` aliasing and the various classifications of types including value types, reference types, pointer types and byref-like types. Essentially this is a different way to achieve "boxing" in order to have a single representation of any parameter.
+
+However, since managed pointers require a reference to a storage location, they does not directly support the same loosely-coupled scenarios as using `object` + boxing for value types. The proposed APIs, however, do make interacting with `object` possible with the new API for the cases that do not require `in\ref\out` variable aliasing for example.
+
+Today, a managed pointer is obtained safely through the `ref` keyword in C#. It references a storage location of an object or value type which can be a stack variable, static variable, parameter, field or array element. If the storage location is a field or array element, the managed pointer is referred to as an "interior pointer" which is supported by GC meaning that GC won't collect the owning object even if there are only interior pointers to it.
+
+A managed pointer can becreated in several ways:
+- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses a `ref byte` approach along with a reference to a `Type`. Note `TypeReference` is a special type and has its own opcodes (mkrefany, refanytype, refanyval) which translate to C# keyworkds (`__makeref`, `__reftype`, `__refvalue`).
```

Comment:
```suggestion
A managed pointer can be created in several ways:
- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses a `ref byte` approach along with a reference to a `Type`. Note `TypeReference` is a special type and has its own opcodes (mkrefany, refanytype, refanyval) which translate to C# keywords (`__makeref`, `__reftype`, `__refvalue`).
```